python/ray/util/sgd/v2/tests/test_trainer.py

```diff
@@ -1,5 +1,126 @@
+from ray.util.sgd.v2 import Trainer, TorchConfig
+from ray.util.sgd.v2.examples.train_linear import train_func as \
+    linear_train_func
+from ray.util.sgd.v2.examples.train_fashion_mnist import train_func as \
+    fashion_mnist_train_func
+
+
+@pytest.fixture
+def ray_start_2_cpus():
+    address_info = ray.init(num_cpus=2)
+    yield address_info
+    # The code after the yield will run as teardown code.
+    ray.shutdown()
+
+
+def test_start_shutdown(ray_start_2_cpus):
+    assert ray.available_resources()["CPU"] == 2
+    trainer = Trainer("torch")
+    trainer.start()
+    time.sleep(1)
+    assert ray.available_resources()["CPU"] == 1
+    trainer.shutdown()
+    time.sleep(1)
+    assert ray.available_resources()["CPU"] == 2
+
+    trainer = Trainer("torch", num_workers=2)
+    trainer.start()
+    time.sleep(1)
+    assert "CPU" not in ray.available_resources()
+    trainer.shutdown()
+    time.sleep(1)
+    assert ray.available_resources()["CPU"] == 2
+
+
+def test_run(ray_start_2_cpus):
+    def train_func(config):
+        return 1
+
+    trainer = Trainer("torch", num_workers=2)
+    trainer.start()
+    results = trainer.run(train_func)
+    trainer.shutdown()
+
+    assert len(results) == 2
+    assert all(result == 1 for result in results)
+
+
+def test_run_config(ray_start_2_cpus):
+    def train_func(config):
+        return config["fruit"]
+
+    config = {"fruit": "banana"}
+
+    trainer = Trainer("torch", num_workers=2)
+    trainer.start()
+    results = trainer.run(train_func, config)
+    trainer.shutdown()
+
+    assert len(results) == 2
+    assert all(result == "banana" for result in results)
+
+
+def test_torch_linear(ray_start_2_cpus):
+    num_workers = 2
+    epochs = 3
+
+    trainer = Trainer(TorchConfig(backend="gloo"), num_workers=num_workers)
+    config = {"lr": 1e-2, "hidden_size": 1, "batch_size": 4, "epochs": epochs}
+    trainer.start()
+    results = trainer.run(linear_train_func, config)
+    trainer.shutdown()
+
+    assert len(results) == num_workers
+
+    for result in results:
+        assert len(result) == epochs
+        assert result[-1]["loss"] < result[0]["loss"]
+
+
+def test_torch_fashion_mnist(ray_start_2_cpus):
+    num_workers = 2
+    epochs = 3
+
+    trainer = Trainer("torch", num_workers=num_workers)
+    config = {"lr": 1e-3, "batch_size": 64, "epochs": epochs}
+    trainer.start()
+    results = trainer.run(fashion_mnist_train_func, config)
+    trainer.shutdown()
+
+    assert len(results) == num_workers
+
+    for result in results:
+        assert len(result) == epochs
+        assert result[-1] < result[0]
+
+
+def test_init_failure(ray_start_2_cpus):
+    with pytest.raises(TypeError):
+        Trainer(5)
+
+    with pytest.raises(ValueError):
+        Trainer("invalid")
+
+
+def test_start_failure(ray_start_2_cpus):
+    with pytest.raises(ValueError):
+        trainer = Trainer("torch", num_workers=0)
+        trainer.start()
+
+
+def test_run_failure(ray_start_2_cpus):
```

Comment:
Do you think we should add another test for worker failure? Just to make sure that if a worker fails, the whole training will error out with a `RuntimeError`.

Reply:
I pushed a fix for this, please take a look

Reply:
Thanks!